File: automation/newsletter/prefs_codec.py
"""Compact serialization of a free-member newsletter filter.

Free (email-only, non-Clerk) subscribers have no Clerk `publicMetadata`
to hold their filter, and Pulpo has no database. Their filter is stored on
the Resend contact's `last_name` field as a side-channel — mirroring the
existing `first_name = "pulpo-locale:<lc>"` hack in subscribers.py.

Why `last_name` and not Resend contact *properties*: the nightly pipeline
reads the whole audience with ONE `GET /audiences/:id/contacts` list call,
and Resend's list response does not include custom properties — reading
properties would force an O(N) per-contact `get`. The list DOES return
`last_name`, which Pulpo never renders (display names come from Clerk), so
it is free to reuse and reads for zero extra API calls.

Format (after the `pulpo-filter:` prefix): `k=v` pairs joined by `;`.
    pulpo-filter:pt=land,house;mx=500000;mn=0;z=el_tunco;cat=beachfront
Keys — all optional, empty ones omitted:
    pt  property_types (csv of slugs)     mx  max_price_usd (int)
    z   zones (csv of slugs)              mn  min_price_usd (int)
    cat categories (csv of slugs)

This module is the READER (pipeline). The WRITER is api/_prefs_codec.js —
the two MUST stay in lockstep; tests/newsletter/test_prefs_codec.py plus the
JS contract test enforce parity. Decode is total: any malformed input
yields an empty dict (→ empty Preference → the cohort fallback), never an
exception.
"""

from __future__ import annotations

from typing import Any

PREFIX = "pulpo-filter:"

# Slug lists we accept. Values outside [a-z0-9_-] are dropped defensively so a
# tampered link can never inject markup or separators back through decode.
_SLUG_KEYS = {"pt": "property_types", "z": "zones", "cat": "categories"}
_INT_KEYS = {"mx": "max_price_usd", "mn": "min_price_usd"}

_ALL_FIELDS = ("property_types", "zones", "categories", "max_price_usd", "min_price_usd")


def _clean_slug(s: str) -> str:
    return "".join(ch for ch in s.strip().lower() if ch.isalnum() or ch in "_-")


def _clean_slug_list(raw: str) -> list[str]:
    out: list[str] = []
    for part in raw.split(","):
        slug = _clean_slug(part)
        if slug and slug not in out:
            out.append(slug)
    return out


def decode(raw: Any) -> dict:
    """Parse a `last_name` value into a filter dict. Total — never raises.

    Returns {} for anything that isn't a well-formed `pulpo-filter:` string,
    which the caller maps to an empty Preference (no opinion).
    """
    if not isinstance(raw, str):
        return {}
    raw = raw.strip()
    if not raw.startswith(PREFIX):
        return {}
    body = raw[len(PREFIX):]
    out: dict = {}
    for pair in body.split(";"):
        if "=" not in pair:
            continue
        k, _, v = pair.partition("=")
        k = k.strip().lower()
        if k in _SLUG_KEYS:
            vals = _clean_slug_list(v)
            if vals:
                out[_SLUG_KEYS[k]] = vals
        elif k in _INT_KEYS:
            digits = "".join(ch for ch in v if ch.isdigit())
            if digits:
                n = int(digits)
                # min=0 carries no information (no floor) — drop it so the
                # encode↔decode round-trip is stable and Preference stays clean.
                if not (k == "mn" and n == 0):
                    out[_INT_KEYS[k]] = float(n)
    return out


def encode(pref: dict) -> str:
    """Serialize a filter dict to a `last_name` value (incl. prefix).

    Empty filter → "" (caller clears the contact's last_name). Kept in sync
    with api/_prefs_codec.js `encode`; the contract test pins the format.
    """
    parts: list[str] = []
    for short, field in _SLUG_KEYS.items():
        vals = pref.get(field)
        if isinstance(vals, list):
            slugs = [s for s in (_clean_slug(str(x)) for x in vals) if s]
            if slugs:
                parts.append(f"{short}={','.join(slugs)}")
    for short, field in _INT_KEYS.items():
        v = pref.get(field)
        if isinstance(v, (int, float)) and v > 0:
            parts.append(f"{short}={int(v)}")
    if not parts:
        return ""
    return PREFIX + ";".join(parts)

File: api/_prefs_codec.js
// Compact serialization of a free-member newsletter filter — the WRITER
// side (the /api/newsletter-prefs endpoint reads the current value to
// prefill the form and writes the new one). The READER is the pipeline's
// automation/newsletter/prefs_codec.py; the two MUST stay in lockstep.
// tests/api/prefs_codec_contract.test.js pins JS↔Python parity on a shared
// fixture set.
//
// The filter lives on the Resend contact's `last_name` field (side-channel),
// NOT contact properties — see prefs_codec.py's header for why (the pipeline
// reads the audience with one list call, which omits properties but returns
// last_name, and Pulpo never renders last_name).
//
// Format (after the `pulpo-filter:` prefix): `k=v` pairs joined by `;`:
//   pulpo-filter:pt=land,house;mx=500000;mn=0;z=el_tunco;cat=beachfront
//   pt property_types (csv)   mx max_price_usd (int)   z zones (csv)
//   mn min_price_usd (int)    cat categories (csv)
//
// decode() is total — malformed input yields {} (→ empty filter), never throws.

const PREFIX = "pulpo-filter:";

const SLUG_KEYS = { pt: "property_types", z: "zones", cat: "categories" };
const INT_KEYS = { mx: "max_price_usd", mn: "min_price_usd" };

function cleanSlug(s) {
  return String(s).trim().toLowerCase().replace(/[^a-z0-9_-]/g, "");
}

function cleanSlugList(raw) {
  const out = [];
  for (const part of String(raw).split(",")) {
    const slug = cleanSlug(part);
    if (slug && !out.includes(slug)) out.push(slug);
  }
  return out;
}

// Parse a `last_name` value into a filter object. Total — never throws.
function decode(raw) {
  if (typeof raw !== "string") return {};
  const s = raw.trim();
  if (!s.startsWith(PREFIX)) return {};
  const body = s.slice(PREFIX.length);
  const out = {};
  for (const pair of body.split(";")) {
    const eq = pair.indexOf("=");
    if (eq < 0) continue;
    const k = pair.slice(0, eq).trim().toLowerCase();
    const v = pair.slice(eq + 1);
    if (SLUG_KEYS[k]) {
      const vals = cleanSlugList(v);
      if (vals.length) out[SLUG_KEYS[k]] = vals;
    } else if (INT_KEYS[k]) {
      const digits = (v.match(/\d/g) || []).join("");
      if (digits) {
        const n = parseInt(digits, 10);
        // min=0 carries no floor — drop it so round-trips are stable.
        if (!(k === "mn" && n === 0)) out[INT_KEYS[k]] = n;
      }
    }
  }
  return out;
}

// Serialize a filter object to a `last_name` value (incl. prefix). Empty
// filter → "" (caller clears last_name). Mirrors prefs_codec.py `encode`.
function encode(pref) {
  const parts = [];
  for (const [short, field] of Object.entries(SLUG_KEYS)) {
    const vals = pref && pref[field];
    if (Array.isArray(vals)) {
      const slugs = vals.map((x) => cleanSlug(x)).filter(Boolean);
      if (slugs.length) parts.push(`${short}=${slugs.join(",")}`);
    }
  }
  for (const [short, field] of Object.entries(INT_KEYS)) {
    const v = pref && pref[field];
    if (typeof v === "number" && Number.isFinite(v) && v > 0) {
      parts.push(`${short}=${Math.trunc(v)}`);
    }
  }
  return parts.length ? PREFIX + parts.join(";") : "";
}

module.exports = { PREFIX, encode, decode };
